File: sample-app-dotnet/src/Publix.Grocery.Api/Inventory/InventoryService.cs
namespace Grocery.Api.Inventory;

// NOTE: Intentionally NOT thread-safe. Module 05 asks learners to reproduce and fix this.
public sealed class InventoryService
{
    private readonly Dictionary<string, int> _onHand = new();

    public int OnHand(string sku) => _onHand.TryGetValue(sku, out var v) ? v : 0;

    public void Receive(string sku, int qty) =>
        _onHand[sku] = OnHand(sku) + qty;

    public bool Adjust(string sku, int delta)
    {
        var current = OnHand(sku);
        if (current + delta < 0) return false;
        _onHand[sku] = current + delta;
        return true;
    }
}

File: sample-app-java/src/main/java/com/example/grocery/inventory/InventoryService.java
package com.example.grocery.inventory;

import java.util.HashMap;
import java.util.Map;

import org.springframework.stereotype.Service;

// NOTE: Intentionally NOT thread-safe. Module 05 asks learners to reproduce and fix this.
@Service
public class InventoryService {
    private final Map<String, Integer> onHand = new HashMap<>();

    public int onHand(String sku) {
        return onHand.getOrDefault(sku, 0);
    }

    public void receive(String sku, int qty) {
        onHand.merge(sku, qty, Integer::sum);
    }

    public boolean adjust(String sku, int delta) {
        int current = onHand.getOrDefault(sku, 0);
        if (current + delta < 0) return false;
        onHand.put(sku, current + delta);
        return true;
    }
}
